### PR TITLE
libobs: Don't keep filter mutex in obs_save_source

### LINKED ISSUE #9893: libobs: deadlock between threads while saving a scene.
### Operating System Info

macOS 14

### Other OS

_No response_

### OBS Studio Version

29.1.3

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

OBS hangs and log file doesn't show this.

### OBS Studio Crash Log URL

_No response_

### Expected Behavior

Saving scene should not cause OBS hangs.

### Current Behavior

deadlock between graphics thread and main thread while saving a scene.

### Steps to Reproduce

I got the callstack here:
```
main thread
0   libsystem_kernel.dylib              0x0000000183baebc8 __psynch_mutexwait + 8
1   libsystem_pthread.dylib             0x0000000183be90c4 _pthread_mutex_firstfit_lock_wait + 84
2   libsystem_pthread.dylib             0x3406000183be6a5c _pthread_mutex_firstfit_lock_slow + 3748683739832516856
3   libobs                              0x777b000105831898 scene_save + 8609475112648835120
4   libobs                              0x0000000105846784 obs_source_save + 180
5   libobs                              0x0000000105851324 obs_save_source + 268
6   libobs                              0x0000000105851654 obs_save_sources_filtered + 164

graphic thread
0   libsystem_kernel.dylib              0x0000000183baebc8 __psynch_mutexwait + 8
1   libsystem_pthread.dylib             0x0000000183be90c4 _pthread_mutex_firstfit_lock_wait + 84
2   libsystem_pthread.dylib             0x6b1a800183be6a5c _pthread_mutex_firstfit_lock_slow + 7717621648941121784
3   libobs                              0x80550001058412a4 obs_source_get_width + 9247297409875181624
4   libobs                              0x0000000105837f30 update_transforms_and_prune_sources + 432
5   libobs                              0x000000010583101c scene_video_render + 96
6   libobs                              0x0000000105840b48 source_render + 316
7   libobs                              0x0000000105841208 obs_source_video_render + 948
```

For better understanding, I made an illustration to show the process, hope it helps.

image

### Anything else we should know?

_No response_

### PR DESCRIPTION
**Pull request by exeldro**
### Description
While filters are saved using obs_save_source other mutex are used. To prevent a deadlock, don't keep the filter mutex while iterating the filters. This fixes #9893

### Motivation and Context
I don't like deadlocks. Always try to keep a mutex locked as short as possible.

### How Has This Been Tested?
On windows 64 tested saving the scene collection.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Review by tt2468 — APPROVED** [review]

**Comment by norihiro** [comment]
A line in the commit message below has 79 characters, where the limit is72:
```text
To prevent a deadlock, don't keep the filter mutex while iterating the filters.
```

**Comment by exeldro** [comment]
rebased and changed commit message

**Comment by walker-WSH** [comment]
how about using `obs_source_get_ref` while saving them into temporary array ?

**Comment by Lain-B** [comment]
I just fixed this to hold references to sources in the list as a safety precaution.

**Review by Lain-B — APPROVED** [review]

File: libobs/obs.c [hunk 1]
```diff
@@ -2563,20 +2564,31 @@ obs_data_t *obs_save_source(obs_source_t *source)
...
+		for (size_t i = filters_copy.num; i > 0; i--) {
+			obs_source_t *filter = filters_copy.array[i - 1];
 			obs_data_t *filter_data = obs_save_source(filter);
 			obs_data_array_push_back(filters, filter_data);
 			obs_data_release(filter_data);
+			obs_source_release(filter);
 		}
 
 		obs_data_set_array(source_data, "filters", filters);
+		da_free(filters_copy);
```
Comment (notr1ch) [inline libobs/obs.c:2589]: This is potentially a memory leak, as `filters_copy` was allocated with `source->filters.num` and not `filters_copy.num`.

Comment (derrod) [reply libobs/obs.c:2589]: I suppose we should either remove the call to `da_reserve()` or make the call to `da_free()` unconditionally.

Comment (Lain-B) [reply libobs/obs.c:2589]: Ah woops, you're right. My apologies for that. Will fix.

Comment (Lain-B) [reply libobs/obs.c:2589]: Fixed in e35824452130320a281f3ed22fe68505fbaf0687 -- sorry for the trouble, and thanks for spotting it.

**Pull request merged**